### PR TITLE
[build] Create separate buildspecs for SM DLCs

### PR DESCRIPTION
**Pull request by saimidu**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right.

### Description
Add separate buildspecs for the latest SM framework version DLCs

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: dlc_developer_config.toml [hunk 1]
```diff
@@ -59,13 +59,18 @@ use_scheduler = false
 # dlc-pr-huggingface-pytorch = "huggingface/pytorch/buildspec-1-9.yml"
 # Setting the buildspec file path to "" allows the image builder to choose the default buildspec file.
 
-# Standard Framework DLC buildspecs
-dlc-pr-autogluon = ""
+# E3 Framework DLC buildspecs
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:62]: nit: Replace `E3` with `E2`

File: dlc_developer_config.toml [hunk 2]
```diff
@@ -59,13 +59,18 @@ use_scheduler = false
...
 dlc-pr-mxnet = ""
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:63]: Should we rename this to `dlc-pr-ec2-pytorch`?

File: mxnet/buildspec-sagemaker.yml [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      source: ../../src/deep_learning_container.py
+      target: deep_learning_container.py
+  inference_context: &INFERENCE_CONTEXT
+    config:
+      source: docker/artifacts/config.properties
+      target: config.properties
+    entrypoint:
+      source: docker/artifacts/mms-entrypoint.py
+      target:  mms-entrypoint.py
+    e3-entrypoint:
```
Comment (tejaschumbalkar) [inline mxnet/buildspec-sagemaker.yml:36]: we do not need `e3` artifacts in the SM buildspecs.

File: mxnet/buildspec.yml [hunk 4]
```diff
@@ -68,33 +68,6 @@ images:
     target: e3
```
Comment (tejaschumbalkar) [inline mxnet/buildspec.yml:68]: Same here, we do not need `sagemaker` artifacts in the EC2 buildspecs.

File: mxnet/buildspec.yml [hunk 5]
```diff
@@ -68,33 +68,6 @@ images:
...
     context:
       <<: *TRAINING_CONTEXT
```
Comment (tejaschumbalkar) [inline mxnet/buildspec.yml:70]: nit: Should we rename the file to add `-ec2` in it so that the usage is clear?

File: pytorch/buildspec.yml [hunk 6]
```diff
@@ -46,139 +46,79 @@ context:
...
-  #   docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile.,
-  #                        *DEVICE_TYPE ]
-  #   target: e3
-  #   context:
-  #     <<: *TRAINING_CONTEXT
-  BuildSageMakerCPUPTTrainPy3DockerImage:
+  BuildE3CPUPTTrainPy3DockerImage:
     <<: *TRAINING_REPOSITORY
     build: &PYTORCH_CPU_TRAINING_PY3 false
+    # enable_test_promotion: false
```
Comment (tejaschumbalkar) [inline pytorch/buildspec.yml:52]: what flag is this?

File: pytorch/buildspec.yml [hunk 7]
```diff
@@ -46,139 +46,79 @@ context:
...
-    # base_image_name: BuildE3CPUPTTrainPy3DockerImage
     device_type: &DEVICE_TYPE cpu
     python_version: &DOCKER_PYTHON_VERSION py3
     tag_python_version: &TAG_PYTHON_VERSION py38
     os_version: &OS_VERSION ubuntu20.04
-    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-sagemaker" ]
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-e3" ]
     docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile., *DEVICE_TYPE ]
-    target: sagemaker
+    target: e3
```
Comment (tejaschumbalkar) [inline pytorch/buildspec.yml:60]: We need to change this to `e2`

**Review by tejaschumbalkar — COMMENTED** [review]
PR needs to be rebased as there is a major change in build code sourced from https://github.com/aws/deep-learning-containers/pull/2140

File: dlc_developer_config.toml (hunk 2 above)
Comment (saimidu) [reply dlc_developer_config.toml:63]: No, the pre-existing CI project names have not been changed. So we currently have `dlc-pr-pytorch` and `dlc-pr-sagemaker-pytorch`.

Comment (saimidu) [reply dlc_developer_config.toml:63]: Do you believe that the older projects should be renamed?

File: dlc_developer_config.toml (hunk 1 above)
Comment (saimidu) [reply dlc_developer_config.toml:62]: Is this E2, or should it be EC2?

File: pytorch/buildspec.yml (hunk 6 above)
Comment (saimidu) [reply pytorch/buildspec.yml:52]: This is a flag that allows contributors to "disable" testing on an image after it is built on CI, with no effect on pipeline tests. It seems to have been "added" by this PR, but it is actually just one of the uncommented lines from the e3 images defined in the buildspec from before this PR.

I can remove this flag in my next revision.

File: mxnet/buildspec.yml (hunk 5 above)
Comment (saimidu) [reply mxnet/buildspec.yml:67]: We could, but it needs an infrastructure level change as well to accompany that.

File: src/start_testbuilds.py [hunk 8]
```diff
@@ -147,7 +156,7 @@ def main():
             if "graviton" in images_str and test_type == "sanity":
                 pr_test_job += "-graviton"
             if is_test_job_enabled(test_type) and is_test_job_implemented_for_framework(images_str, test_type):
-                run_test_job(commit, pr_test_job, images_str)
+                run_test_job(commit, pr_test_job, args.buildspec, images_str)
```
[the remark this answers is not in the document]
Comment (saimidu) [reply src/start_testbuilds.py:159]: The buildspec file needs to be a compulsory input to the `start_testbuilds.py` script, and is therefore passed in through the same method by which we pass it to the `main.py` script.

The arg being passed in is already obtained from the environment variable. Do you recommend defaulting to the environment variable inside the python code in both the `main.py` and the `start_testbuilds.py` cases? Because that change is okay with me.

[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply src/start_testbuilds.py:159]: Defaulting it to a value makes sense to me. It is just my common practice to check any parameter passed to a function before using it within the function.

[the remark this answers is not in the document]
Comment (saimidu) [reply src/start_testbuilds.py:161]: Made the change to default to the value obtained from the environment variable `FRAMEWORK_BUILDSPEC_FILE`.

File: test/test_utils/__init__.py [hunk 9]
```diff
@@ -1535,6 +1540,16 @@ def construct_buildspec_path(dlc_path, framework_path, buildspec, framework_vers
     :param buildspec: buildspec file name
     :param framework_version: default (long) framework version name
     """
+    if get_buildspec_file_path_from_env():
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/test_utils/__init__.py]: Made the recommended change in the next commit. Thanks.

File: dlc_developer_config.toml (hunk 2 above)
Comment (tejaschumbalkar) [reply dlc_developer_config.toml:63]: yes, lets change the project names to be explicitly clear

File: mxnet/buildspec-sagemaker.yml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+      target: mms-entrypoint.py
+    ec2-entrypoint:
+      source: docker/artifacts/mms-entrypoint.py
+      target: mms-ec2-entrypoint.py
+    deep_learning_container:
+      source: ../../src/deep_learning_container.py
+      target: deep_learning_container.py
+
+images:
+  BuildSMMXNetCPUTrainPy3DockerImage:
```
Comment (tejaschumbalkar) [inline mxnet/buildspec-sagemaker.yml:44]: Naming convention seems to be mismatch between PT and TF dockerfiles

File: pytorch/buildspec-sagemaker.yml [hunk 11]
```diff
@@ -0,0 +1,104 @@
+account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
Comment (tejaschumbalkar) [inline pytorch/buildspec-sagemaker.yml:1]: Are we splitting the buildspec for old framework versions as well?

Comment (saimidu) [reply pytorch/buildspec-sagemaker.yml:1]: We can split older versions as needed, but as of now, it only applies for current and future versions.

File: mxnet/buildspec-sagemaker.yml (hunk 10 above)
Comment (saimidu) [reply mxnet/buildspec-sagemaker.yml:44]: I just copy-pasted these from the existing buildspec. Will change these to match the same convention.
